lib/pbench/server/api/resources/query_apis/metadata_index/__init__.py

```diff
@@ -0,0 +1,103 @@
+    subclasses.
+
+    Note that run_id is a required schema parameter for all the
+    classes inheriting this class. Also "preprocess" provides json context
+    containing dataset and a run_id that's passed to the assemble and
+    postprocess methods.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger, schema: Schema):
+        if "run_id" not in schema:
+            raise MissingRunIdSchemaParameter(self.__class__.__name__)
+        super().__init__(config, logger, schema)
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        request is authorized for this dataset.
+
+        If the user has authorization to read the dataset, return the Dataset
+        object and run_id as JSON CONTEXT so that the postprocess operations
+        can use it to identify the index to be searched from document index
+        metadata.
+        """
+        run_id = client_json["run_id"]
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error(f"Dataset with Run ID {run_id!r} not found")
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                f"Dataset owner ID { dataset.owner_id!r} cannot be found in Users"
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # We check authorization against the ownership of the dataset that
+        # was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset.access)
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset, "run_id": run_id}
+
+    def get_index(self, dataset: Dataset, root_index_name: AnyStr) -> AnyStr:
+        """
+        Retrieve the list of ES indices from the metadata table based on a given
+        root_index_name.
+        """
+        try:
+            index_map = Metadata.getvalue(dataset=dataset, key=Metadata.INDEX_MAP)
+        except MetadataError as e:
+            self.logger.error(f"Indices from metadata table not found {e!r}")
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+
+        index_keys = [key for key in index_map if root_index_name in key]
+        if len(index_keys) == 0:
+            self.logger.debug(
+                f"Found no indices matching the prefix {root_index_name}"
+                f"for a dataset {dataset!r}"
+            )
+            abort(HTTPStatus.NOT_FOUND, message="Found no matching indices")
```

Comment:
Do we really want to `abort` in this helper function for a non-error case?  Perhaps return an list of indices and let the API handle this?

Reply:
But `get_index` takes the `root_index_name` as a function parameter to search for that indices, returning a list of all the indices would not be a desirable outcome. Probably `get_index` can return an empty string if it doesn't find any matching indices and then `NOT_FOUND` mechanics can be handled by an API itself rather here if that make sense. 

Also returning an `abort(NOT_FOUND)` seems okay to me since for the end client it wouldn't make much difference whether the server is using an `abort` or `response with 404` internally. 
If we find the matching indices but couldn't find any data for those indices thats the case for returning 200 with an empty json instead of 404.

Reply:
These helpers are closely tied to the API, so I'm not too concerned structurally about the `abort`. In some cases, I created a new exception that would be handled by the base class to `abort`; arguably a bit cleaner, but then `abort` is really just a special exception handled by `Flask`.

And, yes, a comprehensive unit test set for methods like this would be good. We started down the road of using the Flask request framework for API testing, which makes that sort of test really awkward. I started breaking away from that with full true unit tests for the query builder and authorization checker; we should be doubling down on that sort of conversion to make sure our support methods are well tested.

Reply:
Yeah and as Dave pointed out these helper functions are closely tied to the APIs so I guess handling the abort operation here would be okay

Reply:
Yeah we have an issue to reformat the unit tests to move away from request based testing https://github.com/distributed-system-analysis/pbench/issues/2408

Reply:
Not to create a problem, but it does not seem wise to spread API handling behavior outside of the API interface method.  Raising exceptions instead of directly using `abort()` in these cases gives us the ability to see and review how the API responses are handled in one place.

Reply:
okay so if there are no matching indices then this method will return an empty indices string and I moved the empty indices error handling to `assemble`.

Reply:
Can you file a tracking issue to re-factor this code later to bubble `abort()` calls to the top-level API handling function?

Reply:
tracked in https://github.com/distributed-system-analysis/pbench/issues/2529